Hello Eli!

I’m happy to review your project today.
I will mark your mistakes and give you some hints how it is possible to fix them. We are getting ready for real job, where your team leader/senior colleague will do exactly the same. Don't worry and study with pleasure! 

Below you will find my comments - **please do not move, modify or delete them**.

You can find my comments in green, yellow or red boxes like this:

<div class="alert alert-block alert-success">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Success. Everything is done succesfully.
</div>

<div class="alert alert-block alert-warning">
<b>Reviewer's comment</b> <a class="tocSkip"></a>

Remarks. Some recommendations.
</div>

<div class="alert alert-block alert-danger">

<b>Reviewer's comment</b> <a class="tocSkip"></a>

Needs fixing. The block requires some corrections. Work can't be accepted with the red comments.
</div>

You can answer me by using this:

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Thank you so much for the feedback, I appreacaite it! I should have double checked before submitting. Thanks! 
</div>


### Project Introduction:

In this project, I am working for OilyGiant, a mining company, and my task is to find the best location to drill new oil wells to maximize profit. I have data from three regions containing information about potential oil wells, and I need to determine which region will yield the most profitable results while minimizing risks. To do this, I will build a machine learning model to predict oil reserves based on the given data and then evaluate profit and risks using the Bootstrapping technique.

__My Plan for the Project__
 * 1. Data Preparation
     - I will load the datasets from the three regions and check them for missing values and data types.
     - Then, I will split the data for each region into training (75%) and validation (25%) sets.
 * 2. Model Training and Evaluation
     - I will train a Linear Regression model for each region to predict the volume of oil reserves for new wells.
     - After training the model, I will evaluate the predictions on the validation data by calculating the Root Mean Squared Error (RMSE).
     - I will also analyze the average predicted reserves for each region to understand how much oil is expected on average.
 * 3. Profit Calculation Setup
     - I will define key values for calculations:
         - Budget: 100 million
         - Price per barrel: $4.5
         - Number of wells selected per region: 200
     - I will calculate the minimum volume of reserves required to make a profit and compare this to the predicted reserves in each region.
4. Profit Calculation for the Best Wells
     - I will select the top 200 wells with the highest predicted values in each region.
     - Using the predictions, I will calculate the total profit for these selected wells in each region.
     - I will compare the results and identify the region with the highest total profit.
5. Risk and Profit Analysis (Bootstrapping)
     - I will use the Bootstrapping technique with 1000 samples to:
         - Calculate the distribution of profits for each region.
         - Find the average profit and 95% confidence interval for profits.
         - Calculate the risk of losses (negative profit) and ensure it is below 2.5%.
6. Final Recommendation
     - Based on the profit calculations and risk analysis, I will recommend the best region for well development.
     - I will ensure that my recommendation aligns with the business requirements and justifies why the region is the best choice.


### Project Imports:

In [90]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.metrics import mean_squared_error
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.preprocessing import StandardScaler

In [91]:
# Read the datasets for the three regions
region_0 = pd.read_csv('/datasets/geo_data_0.csv')
region_1 = pd.read_csv('/datasets/geo_data_1.csv')
region_2 = pd.read_csv('/datasets/geo_data_2.csv')

In [92]:
# Check the first few rows of each dataset
print('Region 0')
print(region_0.head())
print('--------------')
print('Region 1')
print(region_1.head())
print('--------------')
print('Region 2')
print(region_2.head())

Region 0
      id        f0        f1        f2     product
0  txEyH  0.705745 -0.497823  1.221170  105.280062
1  2acmU  1.334711 -0.340164  4.365080   73.037750
2  409Wp  1.022732  0.151990  1.419926   85.265647
3  iJLyR -0.032172  0.139033  2.978566  168.620776
4  Xdl7t  1.988431  0.155413  4.751769  154.036647
--------------
Region 1
      id         f0         f1        f2     product
0  kBEdx -15.001348  -8.276000 -0.005876    3.179103
1  62mP7  14.272088  -3.475083  0.999183   26.953261
2  vyE1P   6.263187  -5.948386  5.001160  134.766305
3  KcrkZ -13.081196 -11.506057  4.999415  137.945408
4  AHL4O  12.702195  -8.147433  5.004363  134.766305
--------------
Region 2
      id        f0        f1        f2     product
0  fwXo0 -1.146987  0.963328 -0.828965   27.758673
1  WJtFt  0.262778  0.269839 -2.530187   56.069697
2  ovLUW  0.194587  0.289035 -5.586433   62.871910
3  q6cA6  2.236060 -0.553760  0.930038  114.572842
4  WPMUX -0.515993  1.716266  5.899011  149.600746


### Step 1: Data Preparation

First off, let's check for any missing values

In [93]:
# Check for missing values in each dataset
print("Missing values in Region 0:", region_0.isnull().sum().sum())
print("Missing values in Region 1:", region_1.isnull().sum().sum())
print("Missing values in Region 2:", region_2.isnull().sum().sum())

Missing values in Region 0: 0
Missing values in Region 1: 0
Missing values in Region 2: 0


In [94]:
print(region_0.info())
print(region_1.info())
print(region_2.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column   Non-Null Count   Dtype  
---  ------   --------------   -----  
 0   id       100000 non-null  object 
 1   f0       100000 non-null  float64
 2   f1       100000 non-null  float64
 3   f2       100000 non-null  float64
 4   product  100000 non-null  float64
dtypes: float64(4), object(1)
memory usage: 3.8+ MB
None
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 5 columns):
 #   Column  

Based on the information on each regions data there is already good spread and no missing values. So must of the data preparations are already done. 

<div class="alert alert-success">
<b>Reviewer's comment V1</b>

Correct

</div>

Next, we're using StandardScaler during data preparation to make sure all the features in our dataset are on a similar scale. This is important because many machine learning models, including linear regression, perform better when the input data is standardized. When features have different scales, it can skew the model's understanding, leading to less accurate predictions.

By applying StandardScaler, we're centering the data around zero and adjusting the variance so that each feature contributes equally to the model's training. This helps the model learn more effectively and can lead to improved performance.

In [95]:
# Initialize the StandardScaler
scaler = StandardScaler()

# Standardize features for each region
features_0_scaled = scaler.fit_transform(region_0[['f0', 'f1', 'f2']])
features_1_scaled = scaler.fit_transform(region_1[['f0', 'f1', 'f2']])
features_2_scaled = scaler.fit_transform(region_2[['f0', 'f1', 'f2']])

In [96]:
# Select features
features = region_0[['f0', 'f1', 'f2']]

# Initialize StandardScaler and fit_transform the features
scaler = StandardScaler()
scaled_features = scaler.fit_transform(features)

# Convert the scaled features back to a DataFrame
scaled_features_df = pd.DataFrame(scaled_features, columns=features.columns)

# Check the mean and standard deviation
mean = scaled_features_df.mean()
std_dev = scaled_features_df.std()

print("Mean after scaling:\n", mean)
print("Standard deviation after scaling:\n", std_dev)

Mean after scaling:
 f0    1.953993e-17
f1   -5.336176e-17
f2    9.695356e-17
dtype: float64
Standard deviation after scaling:
 f0    1.000005
f1    1.000005
f2    1.000005
dtype: float64


By using StandardScaler, I ensured that all my features have a mean of zero and a standard deviation of one. This way, each feature contributes equally to the model's predictions, preventing any single feature from dominating the results. As a result, my model can more accurately evaluate the potential of each region for oil reserves.

Now lets move on to the model.

<div class="alert alert-success">
<b>Reviewer's comment V1</b>

Good job!

</div>

### Step 2: Training and Testing the model

First off let's split the data:

In [97]:
# Create a list to hold the regions
regions = [region_0, region_1, region_2]

# Prepare to store training and validation sets for each region
train_test_sets = []

for region in regions:
    features = region[['f0', 'f1', 'f2']]
    target = region['product']
    
    # Split the data into training (75%) and validation (25%) sets
    X_train, X_valid, y_train, y_valid = train_test_split(
        features, target, test_size=0.25, random_state=0
    )
    
    # Ensure index alignment between features and targets
    assert X_train.index.equals(y_train.index), "Mismatch between training features and targets"
    assert X_valid.index.equals(y_valid.index), "Mismatch between validation features and targets"
    
    # Store the train and validation sets
    train_test_sets.append((X_train, X_valid, y_train, y_valid))

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

I added a test here to check if there are any mismatches in the index during the splitting of data
</div>

Great! Now let's train the model and make our first prediction:

In [28]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split

# Initialize lists to store predictions and actual values
predictions_list = []
actual_values_list = []

# Loop through each region's data
for idx, region_data in enumerate([region_0, region_1, region_2]):
    print(f"Processing Region {idx}")
    
    # Split the data into features (X) and target (y)
    X = region_data[['f0', 'f1', 'f2']]
    y = region_data['product']
    
    print(f"X shape: {X.shape}, y shape: {y.shape}")  # Debugging shapes

    # Split the data into a training set and validation set at a ratio of 75:25
    X_train, X_valid, y_train, y_valid = train_test_split(X, y, test_size=0.25, random_state=42)

    # Train the model
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    # Make predictions for the validation set
    predictions = model.predict(X_valid)
    
    # Debugging predictions shape
    print(f"Predictions shape for Region {idx}: {predictions.shape}")  # Check shape of predictions
    print(f"Predictions for Region {idx} (first 5): {predictions[:5]}")  # Check first 5 predictions

    # Save the predictions and actual values
    predictions_list.append(predictions)
    actual_values_list.append(y_valid)

# Check lengths after loop
print("Length of predictions_list:", len(predictions_list))
print("Length of actual_values_list:", len(actual_values_list))

Processing Region 0
X shape: (100000, 3), y shape: (100000,)
Predictions shape for Region 0: (25000,)
Predictions for Region 0 (first 5): [101.90101715  78.21777385 115.26690103 105.61861791  97.9801849 ]
Processing Region 1
X shape: (100000, 3), y shape: (100000,)
Predictions shape for Region 1: (25000,)
Predictions for Region 1 (first 5): [ 8.44738063e-01  5.29216119e+01  1.35110385e+02  1.09494863e+02
 -4.72915824e-02]
Processing Region 2
X shape: (100000, 3), y shape: (100000,)
Predictions shape for Region 2: (25000,)
Predictions for Region 2 (first 5): [ 98.30191642 101.59246124  52.4490989  109.92212707  72.41184733]
Length of predictions_list: 3
Length of actual_values_list: 3


Next, let's now save these predictions:

In [29]:
# Prepare to store predictions and actual values
predictions_list = []
actual_values_list = []

# Loop through each region's training and validation set
for i, (X_train, X_valid, y_train, y_valid) in enumerate(train_test_sets):
    # Train the model on the training set
    model = LinearRegression()
    model.fit(X_train, y_train)
    
    # Make predictions for the validation set
    predictions = model.predict(X_valid)
    
    # Store the predictions and actual values
    predictions_list.append(predictions)  # Store predictions
    actual_values_list.append(y_valid)  # Store actual values

    # Print some debug information
    print(f"Region {i} - Predictions: {predictions[:5]}, Actual: {y_valid[:5]}")

Region 0 - Predictions: [ 95.30389144  95.0984074  106.70227293  96.74005877  97.70338055], Actual: 3582      49.723143
60498     10.583795
53227    139.110613
21333     85.109875
3885      39.726827
Name: product, dtype: float64
Region 1 - Predictions: [134.45138055 137.03155518   2.8356238  111.27429673 138.10907187], Actual: 3582     134.766305
60498    137.945408
53227      3.179103
21333    110.992147
3885     137.945408
Name: product, dtype: float64
Region 2 - Predictions: [103.87975876  59.9481585  108.62978239  99.21937684  90.3619444 ], Actual: 3582     129.102262
60498     61.285744
53227     65.026246
21333     62.334048
3885     150.002134
Name: product, dtype: float64


Now we can print the average of the volume of predicted reserves and RMSE:

In [30]:
import numpy as np
from sklearn.metrics import mean_squared_error

# Prepare to store average volumes and RMSEs
average_volumes = []
rmses = []

# Loop through the predictions and actual values
for (predictions, actual_values) in zip(predictions_list, actual_values_list):
    # Calculate the average predicted volume
    average_volume = np.mean(predictions)
    average_volumes.append(average_volume)

    # Calculate RMSE
    rmse = np.sqrt(mean_squared_error(actual_values, predictions))
    rmses.append(rmse)

# Print average volumes and RMSE for each region
for i, (average_volume, rmse) in enumerate(zip(average_volumes, rmses)):
    print(f"Region {i}: Average Volume of Predicted Reserves: {average_volume:.2f}, RMSE: {rmse:.2f}")

Region 0: Average Volume of Predicted Reserves: 92.27, RMSE: 37.48
Region 1: Average Volume of Predicted Reserves: 69.15, RMSE: 0.89
Region 2: Average Volume of Predicted Reserves: 94.71, RMSE: 40.31


Here, we look at the average predicted reserves and the RMSE for each region. This helps us understand how much oil we might expect from each region and how accurate our predictions are.

__Region 0__ shows a good average volume but has a high RMSE, indicating that while the expected reserves are decent, the predictions are not very reliable.

__Region 1__ has a lower average volume but an exceptionally low RMSE, suggesting that although we expect less oil from this region, our predictions are very consistent and trustworthy.

__Region 2__ presents the highest average volume, but like Region 0, it also has a high RMSE, meaning that even though we expect a lot of oil, the predictions can be quite variable and less dependable.

We notice that having a high average volume is great, but it’s crucial to consider the reliability of those predictions too. For instance, even though Region 2 has the highest average volume, its high RMSE raises a flag about the consistency of those predictions.
On the other hand, Region 1, while having the lowest average volume, stands out because its predictions are reliable, which could be more valuable when making investment decisions.

<div class="alert alert-success">
<b>Reviewer's comment V1</b>

Well done!

</div>

### Step 3: Profit Calculations

First, let's store the key values based on the given conditions:

In [31]:
# Set the key values for calculations
budget_per_well = 100_000_000 / 200  # Budget for development of 200 wells
revenue_per_barrel = 4500  # Revenue from one barrel of raw materials
well_capacity = 200  # Number of wells

Next, let's calculate the volume of reserves that will be sufficient for development:

In [32]:
# Calculate the volume of reserves sufficient for developing a new well without losses
target_volume_per_well = budget_per_well / revenue_per_barrel
print("Volume of reserves sufficient for developing a new well:", target_volume_per_well)

Volume of reserves sufficient for developing a new well: 111.11111111111111


In [33]:
# Compare the calculated target volume with the average predicted volumes from each region
average_volumes = {
    'Region 0': 92.27,
    'Region 1': 69.15,
    'Region 2': 94.71,
}

for region, avg_volume in average_volumes.items():
    print(f"{region}: Average Volume: {avg_volume}, Target Volume: {target_volume_per_well}")

Region 0: Average Volume: 92.27, Target Volume: 111.11111111111111
Region 1: Average Volume: 69.15, Target Volume: 111.11111111111111
Region 2: Average Volume: 94.71, Target Volume: 111.11111111111111


In [34]:
# Analyze the findings based on the comparison
for region, avg_volume in average_volumes.items():
    if avg_volume >= target_volume_per_well:
        print(f"{region} has sufficient average volume for development.")
    else:
        print(f"{region} does NOT have sufficient average volume for development.")

Region 0 does NOT have sufficient average volume for development.
Region 1 does NOT have sufficient average volume for development.
Region 2 does NOT have sufficient average volume for development.


The volume of reserves needed to develop a new well is about 111.11 thousand barrels. Now, I'll compare this value with the average predicted reserves for each region:

 * __Region 0__ has an average volume of 92.27 thousand barrels, which is lower than the required amount. This means it's not sufficient for a new well development.

 * __Region 1__ shows an average volume of 69.15 thousand barrels, also falling short of the needed reserves.

 * __Region 2__ has an average volume of 94.71 thousand barrels, which is again below the necessary amount.

I conclude that none of the regions meet the minimum requirement to develop a new well based on the predicted average reserves. However, this information helps us narrow our search. Regions 0 and 2 are relatively close to the required amount, while Region 1 is considerably lower overall.

<div class="alert alert-success">
<b>Reviewer's comment V1</b>

Correct

</div>

### Step 4: Profit Calculation from Selected Oil Wells

Let's make a function that finds the best 200 wells, calculates profit, and summarizes the reserves:

In [71]:
def select_summarize_and_calculate_profit(features, predictions, actual_values, n=500):
    selected = features.copy()
    selected['predictions'] = predictions
    selected['product'] = actual_values
    
    # Select the best wells
    best_wells = selected.sort_values('predictions', ascending=False).head(n)
    
    # Calculate total reserves and profit
    total_reserves = best_wells['product'].sum()
    revenue = total_reserves * 4500
    profit = revenue - 100_000_000
    return best_wells, total_reserves, profit

Now let's use this once on each region:

<div class="alert alert-danger">
<b>Reviewer's comment V1</b>

1. It's better to join all these 3 functions into single one. It doesn't make sense to write these functions separately because you don't need to use the separately:)
2. In the last function you forgot to subtract budget for wells development.

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Does this work?
</div>

<div class="alert alert-success">
<b>Reviewer's comment V2</b>

Yeah, but it's better to use constants instead of hardcoding values manually

</div>

__Region 0__:

In [82]:
# For Region 0
X_valid_0 = train_test_sets[0][1]  # Validation features for Region 0
y_valid_0 = train_test_sets[0][3]  # Actual target values for Region 0
features_df_0 = pd.DataFrame(X_valid_0)

# Select wells, calculate reserves, and compute profit
best_wells_0, total_reserves_0, total_profit_0 = select_summarize_and_calculate_profit(
    features=features_df_0, 
    predictions=predictions_list[0], 
    actual_values=y_valid_0.values
)

# Calculate average values
avg_reserves_0 = total_reserves_0 / len(best_wells_0)
avg_profit_0 = total_profit_0 / len(best_wells_0)

print(f"Region 0: Average Volume of Reserves: {avg_reserves_0:.2f}, Average Profit: {avg_profit_0:.2f}")

Region 0: Average Volume of Reserves: 142.68, Average Profit: 442056.50


__Region 1__:

In [83]:
# For Region 1
X_valid_1 = train_test_sets[1][1]  # Validation features for Region 1
y_valid_1 = train_test_sets[1][3]  # Actual target values for Region 1
features_df_1 = pd.DataFrame(X_valid_1)

# Select wells, calculate reserves, and compute profit
best_wells_1, total_reserves_1, total_profit_1 = select_summarize_and_calculate_profit(
    features=features_df_1, 
    predictions=predictions_list[1], 
    actual_values=y_valid_1.values
)

# Calculate average values
avg_reserves_1 = total_reserves_1 / len(best_wells_1)
avg_profit_1 = total_profit_1 / len(best_wells_1)

print(f"Region 1: Average Volume of Reserves: {avg_reserves_1:.2f}, Average Profit: {avg_profit_1:.2f}")

Region 1: Average Volume of Reserves: 137.94, Average Profit: 420725.72


__Region 2__:

In [84]:
# For Region 2
X_valid_2 = train_test_sets[2][1]  # Validation features for Region 2
y_valid_2 = train_test_sets[2][3]  # Actual target values for Region 2
features_df_2 = pd.DataFrame(X_valid_2)

# Select wells, calculate reserves, and compute profit
best_wells_2, total_reserves_2, total_profit_2 = select_summarize_and_calculate_profit(
    features=features_df_2, 
    predictions=predictions_list[2], 
    actual_values=y_valid_2.values
)

# Calculate average values
avg_reserves_2 = total_reserves_2 / len(best_wells_2)
avg_profit_2 = total_profit_2 / len(best_wells_2)

print(f"Region 2: Average Volume of Reserves: {avg_reserves_2:.2f}, Average Profit: {avg_profit_2:.2f}")

Region 2: Average Volume of Reserves: 134.39, Average Profit: 404776.21


### Step 5: Calculating Risk and Profit for each Region

Let's do some bootstrapping to find the distribution of profit:

In [116]:
def bootstrap_profit(predictions, actual_values, n_wells=500, n_iterations=1000):
    profits = []
    
    for _ in range(n_iterations):
        # Sample 500 indices
        sample_indices = np.random.choice(len(predictions), n_wells, replace=True)
        sample_predictions = predictions[sample_indices]
        sample_actuals = actual_values.iloc[sample_indices]  # Ensure correct indexing

        # Select the best 200 wells based on predictions
        best_indices = np.argsort(sample_predictions)[-200:]  # Get indices of best 200 wells
        selected_predictions = sample_predictions[best_indices]
        selected_actuals = sample_actuals.iloc[best_indices]

        # Calculate profit for selected wells
        best_wells, total_reserves, profit = select_summarize_and_calculate_profit(
            features=pd.DataFrame(selected_actuals),  # Adjust this as needed
            predictions=selected_predictions,
            actual_values=selected_actuals
        )

        profits.append(profit)

    return profits

In [117]:
profits_region_0 = bootstrap_profit(predictions_list[0], y_valid_0)
profits_region_1 = bootstrap_profit(predictions_list[1], y_valid_1)
profits_region_2 = bootstrap_profit(predictions_list[2], y_valid_2)

Now that that is all set up, let's now display the results:

In [119]:
# Display summary statistics for each region
for i, profits in enumerate([profits_region_0, profits_region_1, profits_region_2]):
    total_profit = np.sum(profits)  # Sum all profits for total profit
    avg_profit = np.mean(profits)  # Average profit per iteration
    lower_bound, upper_bound = np.percentile(profits, [2.5, 97.5])  # 95% CI
    
    # Convert profits to a NumPy array for comparison
    profits_array = np.array(profits)
    risk_of_loss = np.mean(profits_array < 0) * 100  # Percentage of negative profits

    print(f"Region {i}:")
    print(f"  Total Profit: {total_profit:.2f}")
    print(f"  Average Profit: {avg_profit:.2f}")
    print(f"  95% Confidence Interval: ({lower_bound:.2f}, {upper_bound:.2f})")
    print(f"  Risk of Loss: {risk_of_loss:.2f}%\n")

Region 0:
  Total Profit: 4374466006.85
  Average Profit: 4374466.01
  95% Confidence Interval: (-932984.70, 9392315.77)
  Risk of Loss: 5.50%

Region 1:
  Total Profit: 4804003212.54
  Average Profit: 4804003.21
  95% Confidence Interval: (793399.02, 8830612.66)
  Risk of Loss: 1.00%

Region 2:
  Total Profit: 4047985496.96
  Average Profit: 4047985.50
  95% Confidence Interval: (-1174194.84, 8947848.29)
  Risk of Loss: 6.30%



After making some changes to how the functions worked the result looks much more trustworthy now. There is slight risk of loss in all three regions now. All of their total profits exceeds the criteria and the confidence intervals have changed quite a bit too. Region 1 is now far ahead of the other two, becoming the clear choice for the best region. Regions 0 and 2 both have confidence intervals in the negatives, and both of their risk of losses do not meet the criteria of the project. Although all their total profits are good enough they aren't as accurate as we need them to be. Which is why now Region 1 is clearly the best region in terms of profit and working well with the model. 

<div class="alert alert-danger">
<b>Reviewer's comment V1</b>

I'm sorry but the results are incorrect. In the correct results the risk in each region is not exact zero. It's a value between 0 and 10%.
    
1. According to the task description you should sample with n=500
2. You have the problem with calculate_profit function which I explained above. 
3. Check the indexes in targets and predictions. They should be the same. Otherwise the results will be incorrect.
4. Check the constants.

</div>

<div class="alert alert-block alert-info">
<b>Student answer.</b> <a class="tocSkip"></a>

Okay now it looks way different after making those changes. The part that was the most confusing for me and still is confusing to me is the sample size. It took me a bit to understand that I needed to train with a sample of 500 and then when I am finalizing the best region then I use the best 200 wells. 
</div>

<div class="alert alert-success">
<b>Reviewer's comment V2</b>

Everything is correct now:) But it's better to use your function `select_summarize_and_calculate_profit` instead of writing a code to calculate the profit inside the loop for the second time.

</div>

### Conclusion:

Based on the data and calculations, it appears that Region 1 is the best option. Its total profit is the highest overall, and has the only positive confidence interval out of all three regions. It's risk of loss is only 1.00 percent which beats the 2.5 percent required for this project. And this is done with the 200 best wells found in my function.

I typically struggle with import statements, but this time I navigated them successfully. I look forward to receiving feedback on this project!